shared/search/providers.ts

```diff
@@ -367,59 +372,96 @@ async function searchReferences(
  *
  * @param search The search function.
  * @param args The arguments to the search function.
+ * @param negateRepoFilter Whether to look only inside or outside the given repo.
  */
 export function searchWithFallback<
-    P extends { repo: string; commit: string; query: string },
+    P extends { repo: string; commit: string; queryTerms: string[] },
     R
 >(
     search: (args: P) => Promise<R>,
     args: P,
     negateRepoFilter = false
 ): Promise<R> {
-    const { repo, commit, query } = args
-    const unindexedQuery = negateRepoFilter
-        ? `${query} -repo:^${repo}$` // commit does not apply to a different repository
-        : `${query} repo:^${repo}$@${commit}`
-    const indexedQuery = negateRepoFilter
-        ? `${query} -repo:^${repo}$ index:only`
-        : `${query} repo:^${repo}$ index:only`
-
     if (getConfig('basicCodeIntel.indexOnly', false)) {
-        return search({
-            ...args,
-            query: indexedQuery,
-        })
+        return searchIndexed(search, args, negateRepoFilter)
     }
 
-    const timeout = getConfig<number>(
-        'basicCodeIntel.unindexedSearchTimeout',
-        5000
-    )
+    // searchIndexed/searchUnindexed modifies queryTerms so make
```

Comment:
Can we make it the responsibility of the callee not to diddle its arguments?

If I ever need to use these functions elsewhere I'm not going to remember this. 🙂